### ISSUE TITLE
client-side throttling resulting in long wait between custom resource creation and sync call

### ISSUE THREAD

**Issue by rouke-broersma** [metacontroller/metacontroller#417]
We have been noticing for a couple months that metacontroller is at times very slow to call our sync hook (up to an hour sometimes) after the custom resource was introduced into the cluster. We see a lot of client-side throttling at those times:

```
{"level":"info","ts":1641464159.2145114,"msg":"Waited for 1.188381736s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/scheduling.k8s.io/v1?timeout=32s\n"}
{"level":"info","ts":1641464179.2140453,"msg":"Waited for 1.194210847s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/autoscaling/v2beta1?timeout=32s\n"}
{"level":"info","ts":1641464199.2149317,"msg":"Waited for 1.192041362s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/coordination.k8s.io/v1beta1?timeout=32s\n"}
{"level":"info","ts":1641464219.214881,"msg":"Waited for 1.195390969s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/authorization.k8s.io/v1?timeout=32s\n"}
{"level":"info","ts":1641464239.2139485,"msg":"Waited for 1.190286683s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/events.k8s.io/v1?timeout=32s\n"}
{"level":"info","ts":1641464259.2137074,"msg":"Waited for 1.193234203s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/autoscaling/v2beta2?timeout=32s\n"}
{"level":"info","ts":1641464279.2138858,"msg":"Waited for 1.189809758s due to client-side throttling, not priority and fairness, request: GET:https://10.0.0.1:443/apis/authorization.k8s.io/v1beta1?timeout=32s\n"}
```

We have recently seen this behavior in some of our own apps that talk to the kubernetes api during runtime and we determined that in our case this was caused by the kubectl cache directory not being writeable by the non-root user running the container.

Can someone clarify if this is expected behavior, a bug or a mistake in our deployment of metacontroller? In any case it is hurting our production environments and users because sync actions are very late.

**Comment by mjsmith1028** [comment]
Hi @rouke-broersma you can increase the values on:

--client-go-qps	Number of queries per second client-go is allowed to make (default 5, e.g. --client-go-qps=100)
--client-go-burst	Allowed burst queries for client-go (default 10, e.g. --client-go-burst=200)

Client-go is handling calls to the k8s api server as well as sync hooks to controllers. These configure the desired behavior for client side throttling.

Also, another argument you may benefit from is:

--workers	Number of sync workers to run (default 5, e.g. --workers=100)

This controls the number of objects that are processed in parallel from the queue.

See https://metacontroller.github.io/metacontroller/guide/configuration.html

Please let me know if this helps or you need any additional info 👍🏻

**Comment by mjsmith1028** [comment]
For reference I am currently configured at:

--client-go-qps=150
--client-go-burst=300
--workers=50

I rarely see any throttling messages with this config.

**Comment by rouke-broersma** [comment]
Hi @mjsmith1028 we deploy metacontroller with helm. How would you suggest we change these settings? There does not appear to be a configuration value for these in the helm chart.

**Comment by mjsmith1028** [comment]
@rouke-broersma The arguments are in the helm chart here: https://github.com/metacontroller/metacontroller/blob/master/deploy/helm/metacontroller/templates/statefulset.yaml#L38

You are right they are not exposed via values.yaml. 

I have a branch with some tweaks to metacontroller and have the statefulset modified as:
```
        {{- with .Values.args }}
        args:
          {{- toYaml . | nindent 10 }}
        {{- end }}
```

Then in values.yaml
```
args:
  - --zap-log-level=4
  - --zap-devel=false
  - --zap-encoder=json
  - --zap-stacktrace-level=error
  - --discovery-interval=1m  # interval for checking if new resources added to cluster, similar to kubectl api-versions
  - --cache-flush-interval=10m  # informer caches k8s object, this flushes the cache and resyncs with k8s api server
  - --metrics-address=:8080  # if 0 will not enable, prometheus metrics address
  - --client-go-qps=150  # queries per second for go rest client
  - --client-go-burst=300  # max queries per second that go rest client can burst to
  - --workers=50  # number of workers that can process the workqueue executing sync hooks in parallel
  - --events-qps=0.0033  # rate at which k8s events can be sent to k8s api server 0.0033 = 5 mins
  - --events-burst=25  # max number of events that can be sent at given rate
```

I think https://github.com/metacontroller/metacontroller/issues/302 will resolve this. @grzesuav  @AmitKumarDas thoughts?

**Comment by grzesuav** [comment]
yes, thanks @mjsmith1028 for explaining. It would be ideally to have a bit refactor the chart and use some third party registry for storing them

**Comment by grzesuav** [comment]
Closing s duplicate of https://github.com/metacontroller/metacontroller/issues/302
